File: Drivers/ItJobs.py
import os
import requests
from dotenv import load_dotenv, set_key
from typing import List
from math import ceil


from Drivers.Job import Job
from Drivers.DriverInterface import DriverInterface


class ItJobs(DriverInterface):
    search = ""
    api_key = ""
    url = ""
    last_published_id = 0

    def __init__(self):
        load_dotenv()

        self.search = os.getenv('ITJOBS_SEARCH')
        self.url = 'http://api.itjobs.pt/job/search.json'
        self.api_key = os.getenv('ITJOBS_API')
        self.last_published_id = int(os.getenv('ITJOBS_LASTPUBLISHEDID'))

    def get(self) -> List[Job]:
        limit = 100
        payload = {'api_key': self.api_key, 'q': self.search, 'limit': 1}
        r = requests.get(self.url, payload)
        total = r.json()['total']
        if (total == 0):
            return []
        pages = ceil(total/limit,)
        currentJobs = []
        for page in range(1, pages + 1):
            payload = {'api_key': self.api_key,
                       'q': self.search, 'limit': limit, 'page': page}
            currentJobs = currentJobs + \
                requests.get(self.url, payload).json()['results']
        currentJobs = list(sorted(currentJobs, key=lambda j: j['id']))
        currentJobs = list(
            filter(lambda j: self.__filterUnpublished(j['id']), currentJobs))

        unpublishedJobs: List[Job] = []
        for j in currentJobs:
            unpublishedJobs.append(Job(j['id'], self.__getJobUrl(j)))
        return unpublishedJobs

    def __filterUnpublished(self, id: int) -> bool:
        return id > self.last_published_id

    def __getJobUrl(self, job) -> str:
        return 'https://www.itjobs.pt/oferta/' + str(job['id']) + '/' + job['slug']

    def persistPublished(self, job: Job):
        if job.identifier > self.last_published_id:
            self.last_published_id = job.identifier
            print('Persisting ITJOBS_LASTPUBLISHEDID==' + str(job.identifier))
            set_key(os.path.join(os.path.dirname(os.path.realpath(
                __file__)), '.env'), 'ITJOBS_LASTPUBLISHEDID', str(job.identifier))

File: src/drivers/ItJobs.js
const axios = require('axios').default
const BaseDriver = require('./BaseDriver')
require('dotenv').config()

module.exports = class ItJobs extends BaseDriver {

    constructor() {
        this.url = 'https://api.itjobs.pt/job/search.json';
        super()
    }

    /**
     * @returns the default config
     */
    defaultConfig() {
        return {
            tags: [],
            last_published_id: 0
        }
    }

    //     * @returns {[{id:number, url:string}], ...} array of job objects, sorted by id
    /**
     * Retrieves jobs from the api
     */
    async getJobs() {
        try {
            console.info("Fetching jobs from ItJobs")
            //nothing to do here
            if (this.config.tags.length === 0) {
                console.info("No tags defined in " + this.configFileName())
                console.info("Returning empty array...")
                return []
            }


            let allJobs = []
            let limit = 10 // limit per request, api imposed
            let tagsList = this.config.tags.length;

            for (let tagConfig = 0; tagConfig < tagsList; tagConfig++) {
                for (let page = 0; page < 1; page++) {
                    console.info('tagConfig ' + this.config.tags[tagConfig])

                    let payload = {
                        'api_key': process.env.TOKEN_ITJOBS,
                        'limit': limit,
                        'q': this.config.tags[tagConfig],
                        'offset': page * limit
                    }

                    let jobs = await axios.get(this.url, {
                        params: payload
                    });

                    if (jobs.data.results.length === 0) {
                        break;
                    } else {
                        allJobs = allJobs.concat(jobs.data.results)
                    }
                }
            }

            console.info('Filtering ...')
            let filteredJobs = allJobs.filter(x => this.filterUnpublished(x.id))


            console.info(filteredJobs.length + ' jobs found ...')

            console.info('Mapping and sorting ...')

            return filteredJobs.map(x => {
                return {
                    id: x.id,
                    url: "https://www.itjobs.pt/oferta/" + x.id + "/" + x.slug

                }
            }).sort(x => x.id)
        } catch (error) {
            console.error("LandingJobs -> getJobs -> error", error)
        }
    }


    /**
     * Filter jobs by job id
     * @param {number} id the job id
     */
    filterUnpublished(id) {
        return parseInt(id) > this.config.last_published_job_id
    }


    storePublishedJob(job) {
        try {
            console.info("Storing job id " + job.id)
            this.config.last_published_job_id = job.id
            this.storeConfig()
        } catch (error) {
            console.error("ItJobs -> storePublishedJob -> error", error)
        }
    }
}